### ISSUE TITLE
Allow re-run of gerrit-based postsubmit at merged state without additional commits.

### ISSUE THREAD

**Issue by briantkennedy** [kubernetes/test-infra#22653]
**What would you like to be added**:
Add a flag to post submit jobs to allow re-running it at the merged state without including later changes.

**Why is this needed**:
The current behavior for re-running a postsubmit is to take the current master, merge it with the patch for the change, then run.  This works well for jobs that need to apply the latest changes to a running system, but is not ideal for artifact generation jobs that get fed into a CI pipeline.  Ff a random failure occurs during artifact generation (expired credentials, etc), then the job needs to get re-run at the merged state, but not include later commits.

**Comment by BenTheElder** [comment]
Is this not is already how postsubmits behave when re-run? I didn't think we re-resolved commits when re-reunning a postsubmit, in fact people want the opposite behavior, when we re-run a postsubmit we just run the same job again with no config updates etc.

**Comment by briantkennedy** [comment]
@chaodaiG Could you please comment regarding the behavior that you explained regarding postsubmit?

**Comment by chaodaiG** [comment]
This issue was from a gerrit based prow instance, what we observed in clone log is:
```
...
$ git checkout master
$ git fetch <REPO_CLONE_URL> <REF_FROM_PR>
From <REPO_CLONE_URL>
 * branch                <REF_FROM_PR> -> FETCH_HEAD
$ git merge --no-ff <GIT_HASH_OF_THE_COMMIT>
...
```

and just noticed that this is different from github based prow, aka prow.k8s.io. In an example postsubmit job https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/post-cluster-api-push-images/1407425707163783168, the clone log is like:
```
# Cloning kubernetes-sigs/cluster-api at master(e57f23aab48fe60bb1304bcf614bc6ea4211a0d5)
...
$ git fetch https://github.com/kubernetes-sigs/cluster-api.git master
From https://github.com/kubernetes-sigs/cluster-api
 * branch                master     -> FETCH_HEAD
$ git checkout e57f23aab48fe60bb1304bcf614bc6ea4211a0d5
...
$ git branch --force master e57f23aab48fe60bb1304bcf614bc6ea4211a0d5
$ git checkout master
...
```

This is a little surprising to me, my immediate feeling is that may be we should make gerrit clone same way as github? @cjwagner @fejta , do you have any comment on this?

**Comment by cjwagner** [comment]
> This is a little surprising to me, my immediate feeling is that may be we should make gerrit clone same way as github? @cjwagner @fejta , do you have any comment on this?

Yeah we should make this match the behavior with GH. I suspect the distinction is that we don't resolve the base branch commit at trigger time and include that specific SHA in the ProwJob Refs like we do for GH. So instead of the base branch SHA being resolved once at trigger time, it is resolved at run time each time it is run.

**Comment by chaodaiG** [comment]
This is fixed by https://github.com/kubernetes/test-infra/pull/22746

**Comment by BenTheElder** [comment]
@chaodaiG shall we close it, or are you still confirming?

**Comment by chaodaiG** [comment]
Not yet confirmed, will need to wait until this is applied on our Gerrit based prow. I will close this one once it's confirmed

**Comment by chaodaiG** [comment]
Unfortunately #22746 didn't quite fix the issue, and reverted in #22866

Reason:
> #22746 changes postsubmit jobs for gerrit PR to clone exactly the SHA from the commit, which could be stale because gerrit ranks commits by commit timestamp, the SHA could be stale, which is not ideal. Will need to figure out something better than this.

Gerrit creates a merge commit right after a PR is submitted, which is exactly what we are looking for here. Unfortunately after some experiments I was not able to find an API to find the associated merge commit from a submitted PR

**Comment by spiffxp** [comment]
/sig testing
yay typos
